google/cloud/storage/client.h

```diff
@@ -561,8 +626,13 @@ class Client {
    *     `IfMetagenerationNotMatch`, `KmsKeyName`, `MD5HashValue`,
    *     `PredefinedAcl`, `Projection`, and `UserProject`.
    *
-   * @throw std::runtime_error if the operation cannot be completed using the
-   *   current policies.
+   * @throw std::runtime_error if there is a permanent failure, or if there were
+   *     more transient failures than allowed by the current retry policy.
+   *
+   * @par Idempotency
+   * Like other operations that mutate the state of GCS, this operation is only
+   * idempotent if restricted by pre-conditions, in this case,
+   * `IfGenerationMatch`.
```

Comment:
Is `IfMetagenerationMatch` also required for an idempotent operation w.r.t Copy? Given metadata can be updated a copy isn't idempotent if the metadata generation changed between two copies?

Reply:
This was unclear to me, writing it down helped me reason through it, please check my math:

- I am going to use the [wikipedia definition](https://en.wikipedia.org/wiki/Idempotence#Computer_science_meaning) of idempotent "the system state remains the same after one or several calls".  The other common definition / rule of thumb is: "the call succeeds at most once".  The second definition implies the first, but not the other way around.

- `Objects: copy` does not change the metageneration, I just tested that. So setting `IfMetagenerationMatch` would not prevent this function from changing the state multiple times.

- `Objects: copy` does change the generation number, I also tested that. So setting `IfGenerationMatch` would prevent the function from succeeding more than once.

- Setting both would also have that effect, but that is unnecessary.

So I think the comment (and the code) is correct.